### PR TITLE
Reorder extensions to put SigAlgs last (master)

### PR DESCRIPTION
**Pull request by tmshort**
WebSphere application server cannot handle having an empty
extension (e.g. EMS/EtM) as the last extension in a client hello.
This moves the SigAlgs extension last for TLSv1.2 to avoid this
issue.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by tmshort** [comment]
Related to #3910, in that this is a "better" fix for the WebSphere issue.
Fixes #2691 for the master branch. A different change is needed for 1.1.0.

**Comment by tmshort** [comment]
Reference:

https://www.ietf.org/mail-archive/web/tls/current/msg19719.html

**Comment by tmshort** [comment]
Please hold this one until I can do further testing.

File: ssl/ssl_locl.h [hunk 1]
```diff
@@ -674,14 +674,14 @@ typedef enum tlsext_index_en {
...
     TLSEXT_IDX_next_proto_neg,
     TLSEXT_IDX_application_layer_protocol_negotiation,
     TLSEXT_IDX_use_srtp,
     TLSEXT_IDX_encrypt_then_mac,
     TLSEXT_IDX_signed_certificate_timestamp,
     TLSEXT_IDX_extended_master_secret,
+    TLSEXT_IDX_signature_algorithms,
     TLSEXT_IDX_supported_versions,
     TLSEXT_IDX_psk_kex_modes,
     TLSEXT_IDX_key_share,
```
Comment (kaduk) [inline ssl/ssl_locl.h]: github won't let me comment that far away, but early_data is also empty in the clienthello.  It's 1.3-only, and you're not supposed to use it against servers that you don't know for sure support 1.3 (and if you use it the PSK extension comes last anyway and is not empty).  I guess that means it doesn't need to move, but it seemed worth noting.

**Comment by tmshort** [comment]
OK, further testing has been done. Looks as though WebSphere (at least as of 8.5.5) cannot handle EtM at all, and it behaves strangely in my testing (continuous sending of . But that would be another commit; SSL_OP_NO_ENCRYPT_THEN_MAC needs to be backported from master.

File: ssl/ssl_locl.h (hunk 1 above)
Comment (tmshort) [inline ssl/ssl_locl.h]: Yes, I ignored early_data precisely for that reason.

**Comment by tmshort** [comment]
Ping @openssl/committers? This is a compatibility fix for WebSphere 7.x

**Comment by snhenson** [comment]
Will this require any changes to the padding extension?

**Comment by kaduk** [comment]
> Will this require any changes to the padding extension?

No.  Despite the title, it's not actually last-last, just past things that might be empty in actual usage.

**Comment by tmshort** [comment]
No, it shouldn't. The padding extension is supposed to be last (except in TLSv1.3). However, the padding extension *can* be zero length, which can be a problem. So, if perhaps that extension never be allowed to be zero length in `tls_construct_ctos_padding()`, then it would be better. But again, the padding extension is user-controlled via SSL_OP_TLSEXT_PADDING

**Comment by davidben** [comment]
When we ran into this way back, we adjusted the length of our padding extension to ensure it's never empty. It's a pretty small change. Just tweak the numbers here slightly.

        if (hlen >= 4)
            hlen -= 4;
        else
            hlen = 0;

**Comment by tmshort** [comment]
@davidben Yes, that's the code I was referencing.

**Comment by tmshort** [comment]
Updated the code referenced by @davidben to have a minimum length of 1.

**Comment by richsalz** [comment]
squashed and merged.  big blue can breathe more easily now ;)  thanks!

**Pull request closed without merging**